### PR TITLE
Log an error when the Build.pk differs from the pk in Build.error message

### PR DESCRIPTION
**Pull request by humitos**
This is my last card played trying to get more debug information to understand what is happening with the Build.pk and the Build.error.

Example: https://readthedocs.org/projects/libmodule/builds/7015103/

### CODE REVIEW

**Comment by humitos** [comment]
@agjohnson please, take a look at this and let me know if this could be a good way to get more debugging information from production environment.

File: readthedocs/builds/models.py [hunk 1]
```diff
@@ -466,6 +466,27 @@ def finished(self):
...
+        msg = 'There was a problem with Read the Docs'
+        if self.error.startswith(msg) and str(self.pk) not in self.error:
+            log.error(
+                'Build.pk differs from "pk" in Build.error',
+                exc_info=True,
+                stack_info=True,
+                extra={
+                    'build': self,
+                    'project': self.project,
+                    'version': self.version,
```
Comment (humitos) [inline readthedocs/builds/models.py]: What else could I add here relevant to this issue?

**Comment by stsewd** [comment]
@humitos there are others examples of this? I didn't find anything with that one :/

**Comment by stsewd** [comment]
Also, that id exists? I think the instance of the `build` dictionary is being shared on calls (something similar to what I experimented here https://github.com/rtfd/readthedocs.org/pull/3764#discussion_r173648808)

**Comment by stsewd** [comment]
From pdb

```
(Pdb) id(self); self.project; self.build
140249780871504
<APIProject: bookpy-2>
{'project': 199282, u'commands': [], u'state_display': u'Triggered', u'success': False, u'state': u'finished', u'setup': u'', u'exit_code': 1, u'output': u'', u'cold_storage': None, u'length': 3
30, 'version': 554, u'id': 367, u'error': 'There was a problem with Read the Docs while building your documentation. Please report this to us with your build id (367).', u'date': u'2018-04-12T00
:56:24.935195', u'commit': u'2c1475517fbfcfe333ed028c27f633313bab65a7', u'setup_error': u'', u'type': u'html', u'builder': u'stsewd'}
(Pdb) c
> /home/stsewd/rtd/readthedocs.org/readthedocs/doc_builder/environments.py(577)update_build()
-> api_v2.build(self.build['id']).put(self.build)
(Pdb) id(self); self.project; self.build
140249780871504
<APIProject: bookpy-2>
{'project': 199282, u'commands': [], u'state_display': u'Triggered', u'success': False, u'state': u'finished', u'setup': u'', u'exit_code': 1, u'output': u'', u'cold_storage': None, u'length': 6
3, 'version': 554, u'id': 367, u'error': 'There was a problem with Read the Docs while building your documentation. Please report this to us with your build id (366).', u'date': u'2018-04-12T00:
56:24.935195', u'commit': u'2c1475517fbfcfe333ed028c27f633313bab65a7', u'setup_error': u'', u'type': u'html', u'builder': u'stsewd'}
```

Here you can see how the message is changed (`error` key), the breakpoints are on 

https://github.com/rtfd/readthedocs.org/blob/64def2d3b73b98c0593185ee901d25091d07f8a0/readthedocs/doc_builder/environments.py#L512

https://github.com/rtfd/readthedocs.org/blob/64def2d3b73b98c0593185ee901d25091d07f8a0/readthedocs/doc_builder/environments.py#L575

I going to keep investigating.

**Comment by stsewd** [comment]
OK, I got it.

The other build id comes from here

https://github.com/rtfd/readthedocs.org/blob/64def2d3b73b98c0593185ee901d25091d07f8a0/readthedocs/projects/tasks.py#L300-L316

Before that the build is shown as failed and with the correct id, then `update_build` gets executed and updates the build with another id.

**Comment by humitos** [comment]
@stsewd thanks for taking a look at this. I'm really confused at the moment :(

I wasn't able to reproduce this in my local instance and I don't really understand the steps you followed to reproduce this in yours. Can you enumerate those steps and how you did the project fail and where the exception has to be raised?

**Comment by agjohnson** [comment]
I'm going to close this. We discovered the underlying issue and don't need this anymore. PR is open at #3946

**Pull request closed without merging**